### PR TITLE
[MI-3650] Added the metric for MS Graph client method time

### PR DESCRIPTION
**Pull request by manojmalik20**
#### Summary
- Created a new package called clientmodels and moved all the structs to that
- Made necessary changes across the project related to the new package
- Created a new timer layer over the client package using a timer_layer.go.tmpl
- Added go generate commands to generate the timer layer over the client package
- Modified the OnActivate hook and GetClient methods to use the timerlayer wrapper over the original client
- Added new histogram metric for measuring the time taken by a client method
- Fixed some of the failing unit tests

#### Ticket Link
None

### CODE REVIEW

File: server/metrics/metrics.go [hunk 1]
```diff
@@ -388,3 +402,9 @@ func (m *metrics) DecrementChangeEventQueueLength(changeType string) {
 		m.changeEventQueueLength.With(prometheus.Labels{"change_type": changeType}).Dec()
 	}
 }
+
+func (m *metrics) ObserveMSGraphClientMethodDuration(method, success string, elapsed float64) {
```
Comment (ayusht2810) [inline server/metrics/metrics.go:406]: https://github.com/mattermost/mattermost-plugin-msteams-sync/pull/354#discussion_r1369014237

File: server/msteams/layer_generators/main.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+func buildTimerLayer() error {
+	code, err := generateLayer("ClientTimerLayer", "timer_layer.go.tmpl")
+	if err != nil {
+		return err
+	}
+
+	formatedCode, err := format.Source(code)
+	if err != nil {
+		log.Println("44")
```
Comment (ayusht2810) [inline server/msteams/layer_generators/main.go:43]:
```suggestion
```

File: server/msteams/layer_generators/main.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+		return err
+	}
+
+	err = os.MkdirAll("client_timerlayer", 0700)
+	if err != nil {
+		log.Println("51")
+		return err
+	}
```
Comment (ayusht2810) [inline server/msteams/layer_generators/main.go:51]:
```suggestion
	if err = os.MkdirAll("client_timerlayer", 0700); err != nil {
		return err
	}
```

Comment (ayusht2810) [reply server/msteams/layer_generators/main.go:51]: also check if we have a constant present for this in golang

File: server/msteams/layer_generators/main.go [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+	return os.WriteFile(path.Join("client_timerlayer", "timerlayer.go"), formatedCode, 0600)
```
Comment (ayusht2810) [inline server/msteams/layer_generators/main.go:53]: same

File: server/msteams/layer_generators/main.go [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+			for _, typeName := range results {
+				if isError(typeName) {
+					return "err == nil"
+				}
+			}
+			return "true"
+		},
+		"errorPresent": func(results []string) bool {
+			for _, typeName := range results {
+				log.Println("188", typeName)
```
Comment (ayusht2810) [inline server/msteams/layer_generators/main.go:184]:
```suggestion
```

File: server/metrics/metrics.go [hunk 6]
```diff
@@ -54,7 +54,10 @@ type Metrics interface {
 	ObserveChangeEventQueueCapacity(count int64)
 	IncrementChangeEventQueueLength(changeType string)
 	DecrementChangeEventQueueLength(changeType string)
+
 	ObserveMSGraphClientMethodDuration(method, success string, elapsed float64)
```
Comment (ayusht2810) [inline server/metrics/metrics.go:58]: club this according to the subsystem.

**Review by ayusht2810 — APPROVED** [review]

**Review by lieut-data — APPROVED** [review]
Thanks, @manojmalik20!

**Review by jespino — APPROVED** [review]
LGTM

**Review by ayusht2810 — APPROVED** [review]

**Pull request merged**